h2o-core/src/main/java/hex/ModelBuilder.java

```diff
@@ -457,6 +456,51 @@ public void run() {
    * @return A F/J Job, which, when executed, does the build.  F/J is NOT started.  */
   abstract protected Driver trainModelImpl();
 
+  private static class Barrier extends CountedCompleter {
+    @Override public void compute() { }
+  }
+
+  /**
+   * Simple wrapper around model task Driver, its main purpose is to make
+   * sure onExceptionalCompletion is not called after join method finishes (similarly how Job behaves).
+   */
+  class TrainModelTaskController {
```

Comment:
just wondering, do you think it is only needed for this type of task? or could it be useful as a more generic abstraction?
Rephrased: are all other tasks wrapped into a Job, and therefore they don't need this? My concern is that if we need an ad-hoc wrapper here, some other `MRTask` or `H2OCountedCompleter` instances may have a similar need.

Reply:
Agreed, might be needed in other places as well. 

MRTask should be fine, it has lots of special handling and `onExceptionalCompletion` is declared final. The only other place where I see a possible issue is FrameSplitter, however, that one I only see used in tests.